server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxMetaData;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.model.MailboxQuery;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.mail.MailboxMapperFactory;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.apache.james.mailbox.store.mail.model.MailboxId;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.apache.james.mailbox.store.mail.model.impl.PropertyBuilder;
+import org.apache.james.mailbox.store.mail.model.impl.SimpleMailboxMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.functions.ThrowingFunction;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+
+public class SetMessagesCreationProcessor<Id extends MailboxId> implements SetMessagesProcessor<Id> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SetMessagesCreationProcessor.class);
+
+    private final MailboxMapperFactory<Id> mailboxMapperFactory;
+    private final MailboxManager mailboxManager;
+    private final MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory;
+    private final MIMEMessageConverter mimeMessageConverter;
+
+    @Inject
+    @VisibleForTesting
+    SetMessagesCreationProcessor(MailboxMapperFactory<Id> mailboxMapperFactory,
+                                 MailboxManager mailboxManager,
+                                 MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory, MIMEMessageConverter mimeMessageConverter) {
+        this.mailboxMapperFactory = mailboxMapperFactory;
+        this.mailboxManager = mailboxManager;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.mimeMessageConverter = mimeMessageConverter;
+    }
+
+    @Override
+    public SetMessagesResponse process(SetMessagesRequest request, MailboxSession mailboxSession) {
+        Mailbox outbox;
+        try {
+            outbox = getOutbox(mailboxSession).orElseThrow(() -> new MailboxRoleNotFoundException(Role.OUTBOX));
+        } catch (MailboxException | MailboxRoleNotFoundException e) {
+            LOGGER.error("Unable to find a mailbox with role 'outbox'!");
+            throw Throwables.propagate(e);
+        }
+        return request.getCreate().entrySet().stream()
+                .map(e -> new MessageWithId.CreationMessageEntry(e.getKey(), e.getValue()))
+                .map(nuMsg -> createMessageInOutbox(nuMsg, mailboxSession, outbox, buildMessageIdFunc(mailboxSession, outbox)))
+                .reduce(SetMessagesResponse.builder(),
+                        (builder, msg) -> builder.created(ImmutableMap.of(msg.creationId, msg.message)),
```

Comment:
You may define the accumulator & the combiner in SetMessagesResponse, as they are dupplicate in each processor.